File: rust/src/day5.rs
use std::collections::HashMap;

use crate::util;

#[derive(Debug)]
struct Line(i32, i32, i32, i32);

fn parse_input(input: &str) -> Vec<Line> {
    input
        .trim()
        .lines()
        .map(|l| {
            let mut segs = l.split(" -> ");
            let mut start = segs.next().unwrap().split(',');
            let mut end = segs.next().unwrap().split(',');

            let x1 = util::parse_i32(start.next().unwrap());
            let y1 = util::parse_i32(start.next().unwrap());
            let x2 = util::parse_i32(end.next().unwrap());
            let y2 = util::parse_i32(end.next().unwrap());

            Line(x1, y1, x2, y2)
        })
        .collect()
}

pub fn order(a: i32, b: i32) -> (i32, i32) {
    if a > b {
        (b, a)
    } else {
        (a, b)
    }
}

pub fn part1(input: &str) -> u64 {
    let mut map = HashMap::<(i32, i32), i32>::new();

    for Line(x1, y1, x2, y2) in parse_input(input) {
        if y1 == y2 {
            let (x1, x2) = order(x1, x2);
            for x in x1..(x2 + 1) {
                *map.entry((x, y1)).or_default() += 1;
            }
        } else if x1 == x2 {
            let (y1, y2) = order(y1, y2);
            for y in y1..(y2 + 1) {
                *map.entry((x1, y)).or_default() += 1;
            }
        }
    }

    map.values().filter(|n| **n >= 2).count() as u64
}

pub fn part2(input: &str) -> u64 {
    let mut map = HashMap::<(i32, i32), i32>::new();

    for Line(x1, y1, x2, y2) in parse_input(input) {
        let dx = x2 - x1;
        let dy = y2 - y1;
        let d = dx.abs().max(dy.abs());
        let mx = dx.signum();
        let my = dy.signum();

        for i in 0..(d + 1) {
            let cx = x1 + mx * i;
            let cy = y1 + my * i;
            *map.entry((cx, cy)).or_default() += 1;
        }
    }

    map.values().filter(|n| **n >= 2).count() as u64
}

#[cfg(test)]
mod tests {
    use super::*;

    const TEST_INPUT: &str = r#"
0,9 -> 5,9
8,0 -> 0,8
9,4 -> 3,4
2,2 -> 2,1
7,0 -> 7,4
6,4 -> 2,0
0,9 -> 2,9
3,4 -> 1,4
0,0 -> 8,8
5,5 -> 8,2
    "#;

    #[test]
    fn test_part1() {
        assert_eq!(part1(TEST_INPUT), 5);
    }

    #[test]
    fn test_part2() {
        assert_eq!(part2(TEST_INPUT), 12);
    }
}

File: zig/src/day5.zig
const std = @import("std");
const math = std.math;
const util = @import("util.zig");

const Map = struct {
    data: std.AutoHashMap(Point, i32),

    const Self = @This();

    const Point = struct { x: i32, y: i32 };

    fn init() Self {
        return .{
            .data = std.AutoHashMap(Point, i32).init(std.heap.page_allocator),
        };
    }

    fn deinit(self: *Self) void {
        self.data.deinit();
    }

    fn mark(self: *Self, x: i32, y: i32) void {
        const res = self.data.getOrPut(Point{ .x = x, .y = y }) catch unreachable;
        if (res.found_existing) {
            res.value_ptr.* += 1;
        } else {
            res.value_ptr.* = 1;
        }
    }

    fn get(self: *const Self, x: i32, y: i32) ?i32 {
        return self.data.get(Point{ .x = x, .y = y });
    }

    fn count(self: *const Self) u64 {
        var total: u64 = 0;
        var values = self.data.valueIterator();
        while (values.next()) |v| {
            if (v.* >= 2) total += 1;
        }
        return total;
    }
};

pub fn part1(input: []const u8) u64 {
    var map = Map.init();
    defer map.deinit();

    var lines = std.mem.tokenize(u8, input, "\n");
    while (lines.next()) |line| {
        var parts = std.mem.tokenize(u8, line, " -> ");
        var start = std.mem.tokenize(u8, parts.next().?, ",");
        var end = std.mem.tokenize(u8, parts.next().?, ",");

        const x1 = util.parseInt(i32, start.next().?);
        const y1 = util.parseInt(i32, start.next().?);
        const x2 = util.parseInt(i32, end.next().?);
        const y2 = util.parseInt(i32, end.next().?);

        if (x1 == x2) {
            const min = std.math.min(y1, y2);
            const max = std.math.max(y1, y2);

            var y = min;
            while (y <= max) : (y += 1) {
                map.mark(x1, y);
            }
        } else if (y1 == y2) {
            const min = math.min(x1, x2);
            const max = math.max(x1, x2);

            var x = min;
            while (x <= max) : (x += 1) {
                map.mark(x, y1);
            }
        }
    }

    //var y: i32 = 0;
    //while (y < 10) : (y += 1) {
    //    var x: i32 = 0;
    //    while (x < 10) : (x += 1) {
    //        if (map.get(x, y)) |v| {
    //            std.debug.print("{}", .{v});
    //        } else {
    //            std.debug.print(".", .{});
    //        }
    //    }
    //    std.debug.print("\n", .{});
    //}

    return map.count();
}

fn signum(n: i32) i32 {
    if (n < 0) {
        return -1;
    } else if (n == 0) {
        return 0;
    } else {
        return 1;
    }
}

fn abs(n: i32) i32 {
    if (n < 0) {
        return -n;
    } else {
        return n;
    }
}

pub fn part2(input: []const u8) u64 {
    var map = Map.init();
    defer map.deinit();

    var lines = std.mem.tokenize(u8, input, "\n");
    while (lines.next()) |line| {
        var parts = std.mem.tokenize(u8, line, " -> ");
        var start = std.mem.tokenize(u8, parts.next().?, ",");
        var end = std.mem.tokenize(u8, parts.next().?, ",");

        const x1 = util.parseInt(i32, start.next().?);
        const y1 = util.parseInt(i32, start.next().?);
        const x2 = util.parseInt(i32, end.next().?);
        const y2 = util.parseInt(i32, end.next().?);

        const d = math.max(abs(x2 - x1), abs(y2 - y1));
        const mx = signum(x2 - x1);
        const my = signum(y2 - y1);

        var c: i32 = 0;
        while (c <= d) : (c += 1) {
            const cx = x1 + c * mx;
            const cy = y1 + c * my;
            map.mark(cx, cy);
        }
    }

    return map.count();
}

test "day 5 part 1" {
    const test_input =
        \\0,9 -> 5,9
        \\8,0 -> 0,8
        \\9,4 -> 3,4
        \\2,2 -> 2,1
        \\7,0 -> 7,4
        \\6,4 -> 2,0
        \\0,9 -> 2,9
        \\3,4 -> 1,4
        \\0,0 -> 8,8
        \\5,5 -> 8,2
    ;
    try std.testing.expectEqual(part1(test_input), 5);
}

test "day 5 part 2" {
    const test_input =
        \\0,9 -> 5,9
        \\8,0 -> 0,8
        \\9,4 -> 3,4
        \\2,2 -> 2,1
        \\7,0 -> 7,4
        \\6,4 -> 2,0
        \\0,9 -> 2,9
        \\3,4 -> 1,4
        \\0,0 -> 8,8
        \\5,5 -> 8,2
    ;
    try std.testing.expectEqual(part2(test_input), 12);
}
